### ISSUE TITLE
supply buffer to read function

### ISSUE THREAD

**Issue by JimHokanson** [kuba--/zip#35]
I'm working with a memory managed system (Matlab) and was wondering if it would be possible to have the following functions:

1. extern size_t *zip_entry_size(struct zip_t *zip) - return the # of bytes of the current entry
2. extern int zip_entry_read2(struct zip_t *zip, void *buf, size_t bufsize) -User passes in the buffer, the bufsize here is the size of the input buffer ...

The alternative, which is ok for now, is to memcpy after reading, but I'd prefer to avoid this.

**Comment by kuba--** [comment]
@JimHokanson - regarding the first function it's not a problem (I also thought about more generic function `zip_entry_stats` which will return more details like compressed and uncompressed size etc.).
Regarding the `zip_entry_read2`. We have `extern int zip_entry_read(struct zip_t *zip, void **buf, size_t *bufsize);` which allocates buffer for you. So, if I understand correctly you want to allocate buffer by yourself, so the function will read no more than given size of the buffer. Something like `snprintf`.  We can call it `zip_entry_nread`.

**Comment by JimHokanson** [comment]
Great suggestion with the stats!

I hadn't been thinking the same way as you with `zip_entry_nread`, but that seems useful as well. This of course could be generalized to reading from start to stop bytes (or start and length).

Anything I return to the user has to come from managed memory, so my current code would need to be something like:

```
zip_entry_read(zip, &buf, &bufsize);
//Allocation of managed memory
uint8_t *data = (uint8_t *) mxMalloc(bufsize);
memcpy(data,buf,bufsize); 
free(buf)
``` 

This isn't the end of the world although it does double my memory requirements.

**Comment by kuba--** [comment]
@JimHokanson check this: https://github.com/kuba--/zip/pull/37 out
I've decided to add two new functions:
```
extern unsigned long long zip_entry_size(struct zip_t *zip);
extern unsigned int zip_entry_crc32(struct zip_t *zip);
```
to be consistent with what we have so far (`zip_entry_name, zip_entry_index, zip_entry_isdir, ...`)

**Comment by JimHokanson** [comment]
Great! At this point what are your thoughts on the second part regarding having the user allocate the memory?

**Comment by kuba--** [comment]
@JimHokanson - the second part will come in next PR, soon.
Most likely the signature will be sth. like:
```
extern int zip_entry_nread(struct zip_t *zip, void *buf, size_t n);
```
or 
```
extern int zip_entry_noallocread(struct zip_t *zip, void *buf, size_t bufsize);
```

**Comment by JimHokanson** [comment]
Either would be fine. My first instinct was for the second but I think the first could be useful. The first however conflates the functionality of no allocation with the feature of only reading a subset of the data. So for example, why can't I read n bytes that the function allocates.

Also to generalize more, I was thinking it would be better to specify a start and length, rather than just the number of bytes. That being said, I'm not sure what the use case for such functionality would be! On the other hand perhaps being able to peek at the first n bytes of a file would be informative.

So in summary, either would be fine and would allow me to accomplish what I want to. I know designing these interfaces can be challenging and I appreciate the added functionality (as well as the repo itself).

**Comment by kuba--** [comment]
PR: https://github.com/kuba--/zip/pull/38
```
/*
  Extracts the current zip entry into a memory buffer using no memory allocation.

  Args:
    zip: zip archive handler.
    buf: preallocated output buffer.
    bufsize: buffer size (in bytes).

  Note:
    - ensure supplied output buffer is large enough.
    - zip_entry_size returns uncompressed size for the current zip entry.

  Returns:
    The return code - 0 on success, negative number (< 0) on error (e.g. bufsize
    is not large enought).
*/
extern int zip_entry_noallocread(struct zip_t *zip, void *buf, size_t bufsize);
```

**Comment by kuba--** [comment]
@JimHokanson - it's already been merged into the master

**Comment by kuba--** [comment]
@JimHokanson - I hope it works for you. Later we can extend API

**Comment by kuba--** [comment]
@JimHokanson - it's fixed (sorry about that, I forgot about recently added `zip_entry_openbyindex`). 
I also fixed the example (it was copy-paste from test). Yep, for binary data extra NULL does not matter.

**Comment by JimHokanson** [comment]
Yeah, sorry it took me a bit to test it. I can confirm the functionality now works as expected, thanks.
